File: app/src/main/java/com/lucagiorgetti/surprix/ui/loginfragments/home/LoginHomeFragment.java
package com.lucagiorgetti.surprix.ui.loginfragments.home;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.navigation.Navigation;

import com.facebook.CallbackManager;
import com.facebook.appevents.AppEventsLogger;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;
import com.lucagiorgetti.surprix.R;
import com.lucagiorgetti.surprix.SurprixApplication;
import com.lucagiorgetti.surprix.listenerInterfaces.CallbackWithExceptionInterface;
import com.lucagiorgetti.surprix.ui.activities.MainActivity;
import com.lucagiorgetti.surprix.utility.BaseFragment;
import com.lucagiorgetti.surprix.utility.LoginFlowHelper;
import com.lucagiorgetti.surprix.utility.SystemUtils;

import timber.log.Timber;

public class LoginHomeFragment extends BaseFragment {

    private CallbackManager callbackManager;
    private GoogleSignInClient googleSignInClient;
    public static final int RC_SIGN_IN = 2345;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN).requestIdToken(getString(R.string.default_web_client_id)).requestEmail().build();
        googleSignInClient = GoogleSignIn.getClient(getActivity(), gso);
        AppEventsLogger.activateApp(SurprixApplication.getInstance());
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_login_home, container, false);

        callbackManager = CallbackManager.Factory.create();

        ProgressBar progressBar = root.findViewById(R.id.login_loading);
        setProgressBar(progressBar);

        Button facebookCustomLogin = root.findViewById(R.id.btn_start_facebook);
        Button login = root.findViewById(R.id.btn_sign_in);
        Button registerBtn = root.findViewById(R.id.btn_sign_up);
        Button googleLogin = root.findViewById(R.id.btn_google_login);

        facebookCustomLogin.setOnClickListener(view -> {
            if (SystemUtils.checkNetworkAvailability()) {
                LoginFlowHelper.loginWithFacebook(getActivity(), this, callbackManager, new CallbackWithExceptionInterface() {
                    @Override
                    public void onStart() {
                        showLoading();
                    }

                    @Override
                    public void onSuccess() {
                        hideLoading();
                        SystemUtils.openNewActivityWithFinishing(getActivity(), MainActivity.class);
                       // getActivity().finish();
                    }

                    @Override
                    public void onFailure(Exception e) {
                        hideLoading();
                        if (e instanceof LoginFlowHelper.UserNeedToCompleteSignUpException) {
                            hideLoading();
                            Navigation.findNavController(view).navigate(LoginHomeFragmentDirections.actionNavigationLoginHomeToNavigationLoginPrivacy(((LoginFlowHelper.UserNeedToCompleteSignUpException) e).getCurrentUser().getEmail(), true));
                        } else {
                            Toast.makeText(getContext(), e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    }
                });
            }
        });

        googleLogin.setOnClickListener(view -> {
            if (SystemUtils.checkNetworkAvailability()) {
                Intent signInIntent = googleSignInClient.getSignInIntent();
                showLoading();
                startActivityForResult(signInIntent, RC_SIGN_IN);
            }
        });

        login.setOnClickListener(v -> Navigation.findNavController(v).navigate(LoginHomeFragmentDirections.actionNavigationLoginHomeToNavigationLoginSignin()));
        registerBtn.setOnClickListener(v -> Navigation.findNavController(v).navigate(LoginHomeFragmentDirections.actionNavigationLoginHomeToNavigationLoginPrivacy(null, false)));
        return root;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == RC_SIGN_IN){
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            try {
                // Google Sign In was successful, authenticate with Firebase
                GoogleSignInAccount account = task.getResult(ApiException.class);
                Timber.d("firebaseAuthWithGoogle:%s", account.getId());
                LoginFlowHelper.loginWithGoogle(getActivity(), account.getIdToken(), new CallbackWithExceptionInterface() {
                    @Override
                    public void onStart() {

                    }

                    @Override
                    public void onSuccess() {
                        hideLoading();
                        SystemUtils.openNewActivityWithFinishing(getActivity(), MainActivity.class);
                    }

                    @Override
                    public void onFailure(Exception e) {
                        if (e instanceof LoginFlowHelper.UserNeedToCompleteSignUpException) {
                            hideLoading();
                            Navigation.findNavController(getView()).navigate(LoginHomeFragmentDirections.actionNavigationLoginHomeToNavigationLoginPrivacy(((LoginFlowHelper.UserNeedToCompleteSignUpException) e).getCurrentUser().getEmail(), true));
                        } else {
                            Toast.makeText(getContext(), e.getMessage(), Toast.LENGTH_SHORT).show();
                        }                    }
                });
            } catch (ApiException e) {
                // Google Sign In failed, update UI appropriately
                Timber.w("Google sign in failed");
                // ...
            }


        } else {
            try {
                callbackManager.onActivityResult(requestCode, resultCode, data);
            } catch (Exception e){
                System.out.println(e.getLocalizedMessage());
            }
        }
    }

    @Override
    public void onResume() {
        super.onResume();
    }
}

File: app/src/main/java/com/lucagiorgetti/surprix/ui/loginfragments/home/LoginHomeFragment.kt
package com.lucagiorgetti.surprix.ui.loginfragments.home

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.credentials.CredentialManager
import androidx.credentials.CustomCredential
import androidx.credentials.GetCredentialRequest
import androidx.credentials.exceptions.GetCredentialCancellationException
import androidx.navigation.Navigation.findNavController
import com.facebook.CallbackManager
import com.facebook.CallbackManager.Factory.create
import com.facebook.appevents.AppEventsLogger
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.libraries.identity.googleid.GetGoogleIdOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import com.google.firebase.auth.AuthResult
import com.lucagiorgetti.surprix.R
import com.lucagiorgetti.surprix.SurprixApplication
import com.lucagiorgetti.surprix.listenerInterfaces.LoginFlowCallbackInterface
import com.lucagiorgetti.surprix.ui.activities.MainActivity
import com.lucagiorgetti.surprix.utility.BaseFragment
import com.lucagiorgetti.surprix.utility.LoginFlowHelper
import com.lucagiorgetti.surprix.utility.LoginFlowHelper.UserNeedToCompleteSignUpException
import com.lucagiorgetti.surprix.utility.SystemUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.launch
import timber.log.Timber
import java.security.MessageDigest
import java.util.UUID
import kotlin.coroutines.CoroutineContext

class LoginHomeFragment : BaseFragment(), CoroutineScope {
    private var callbackManager: CallbackManager? = null
    private var googleSignInClient: GoogleSignInClient? = null
    private var job: Job = Job()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN).requestIdToken(getString(R.string.default_web_client_id)).requestEmail().build()
        googleSignInClient = GoogleSignIn.getClient(requireActivity(), gso)
        AppEventsLogger.activateApp(SurprixApplication.instance)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.fragment_login_home, container, false)
        callbackManager = create()
        val progressBar = root.findViewById<ProgressBar>(R.id.login_loading)
        setProgressBar(progressBar)
        val facebookCustomLogin = root.findViewById<Button>(R.id.btn_start_facebook)
        val login = root.findViewById<Button>(R.id.btn_sign_in)
        val registerBtn = root.findViewById<Button>(R.id.btn_sign_up)
        val googleLogin = root.findViewById<Button>(R.id.btn_google_login)

        facebookCustomLogin.setOnClickListener { view: View? ->
            if (SystemUtils.checkNetworkAvailability()) {
                LoginFlowHelper.loginWithFacebook(requireActivity(), this, callbackManager, object : LoginFlowCallbackInterface {
                    override fun onStart() {
                        showLoading()
                    }

                    override fun onSuccess() {
                        hideLoading()
                        SystemUtils.openNewActivityWithFinishing(requireActivity(), MainActivity::class.java)
                        // getActivity().finish();
                    }

                    override fun onFailure(e: Exception) {
                        hideLoading()
                        if (e is UserNeedToCompleteSignUpException) {
                            hideLoading()
                            findNavController(view!!).navigate(LoginHomeFragmentDirections.actionNavigationLoginHomeToNavigationLoginPrivacy(e.currentUser?.email, true))
                        } else {
                            Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
                        }
                    }
                })
            }
        }
        googleLogin.setOnClickListener {
            if (SystemUtils.checkNetworkAvailability()) {
                showLoading()
                launch {
                    googleSignIn(requireContext())
                }
            } else {
                hideLoading()
            }
        }
        login.setOnClickListener { v: View? -> findNavController(v!!).navigate(LoginHomeFragmentDirections.actionNavigationLoginHomeToNavigationLoginSignin()) }
        registerBtn.setOnClickListener { v: View? -> findNavController(v!!).navigate(LoginHomeFragmentDirections.actionNavigationLoginHomeToNavigationLoginPrivacy(null, false)) }
        return root
    }

    private suspend fun googleSignIn(context: Context) {
        try {
            // Initialize Credential Manager
            val credentialManager: CredentialManager = CredentialManager.create(context)

            // Generate a nonce (a random number used once)
            val ranNonce: String = UUID.randomUUID().toString()
            val bytes: ByteArray = ranNonce.toByteArray()
            val md: MessageDigest = MessageDigest.getInstance("SHA-256")
            val digest: ByteArray = md.digest(bytes)
            val hashedNonce: String = digest.fold("") { str, it -> str + "%02x".format(it) }

            // Set up Google ID option
            val googleIdOption: GetGoogleIdOption = GetGoogleIdOption.Builder()
                .setFilterByAuthorizedAccounts(false)
                .setServerClientId("500151287055-r40li7rdm5ta8illnlpgjjv7rbqfh59a.apps.googleusercontent.com")
                .setNonce(hashedNonce)
                .build()

            // Request credentials
            val request: GetCredentialRequest = GetCredentialRequest.Builder()
                .addCredentialOption(googleIdOption)
                .build()

            // Get the credential result
            val result = credentialManager.getCredential(context, request)
            val credential = result.credential

            // Check if the received credential is a valid Google ID Token
            if (credential is CustomCredential && credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
                val googleIdTokenCredential =
                    GoogleIdTokenCredential.createFrom(credential.data)

                Timber.d("firebaseAuthWithGoogle:%s", googleIdTokenCredential.id)
                LoginFlowHelper.loginWithGoogle(requireActivity(), googleIdTokenCredential.idToken, object : LoginFlowCallbackInterface {
                    override fun onStart() {}
                    override fun onSuccess() {
                        hideLoading()
                        SystemUtils.openNewActivityWithFinishing(requireActivity(), MainActivity::class.java)
                    }

                    override fun onFailure(e: Exception) {
                        if (e is UserNeedToCompleteSignUpException) {
                            hideLoading()
                            findNavController(view!!).navigate(LoginHomeFragmentDirections.actionNavigationLoginHomeToNavigationLoginPrivacy(e.currentUser?.email, true))
                        } else {
                            Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
                        }
                    }
                })
            } else {
                throw RuntimeException("Received an invalid credential type")
            }
        } catch (e: GetCredentialCancellationException) {
            Timber.e(Exception("Sign-in was canceled. Please try again."))
        } catch (e: Exception) {
            Timber.e(e)
        }
    }

    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job
}
